staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -239,12 +628,48 @@ func waitForAdded(q DelayingInterface, depth int) error {
 	})
 }
 
-func waitForWaitingQueueToFill(q DelayingInterface) error {
-	return wait.Poll(1*time.Millisecond, 10*time.Second, func() (done bool, err error) {
-		if len(q.(*delayingType).waitingForAddCh) == 0 {
-			return true, nil
+// QueueState provides a convenient way to confirm the both the Active and Waiting parts of the delaying queue
+// are in the correct state
+type QueueState struct {
+	active  []interface{}
+	waiting []interface{}
+}
+
+// assertQueueState allows us to assert exactly what items we expect to be in the 'active' and 'waiting' queues.
+// This is more declarative and thorough than observing all of the behaviours from the 'active' queue alone.
+func AssertDelayedQueueState(t *testing.T, q EnhancedDelayingInterface, qs QueueState) {
+	var errored bool
+
+	// check the 'waiting' queue
+	wlen := q.LenWaiting()
+	if wlen != len(qs.waiting) {
+		t.Errorf("the waiting queue has %d queued items but we expected %d", wlen, len(qs.waiting))
+		errored = true
+	}
+	for _, w := range qs.waiting {
+		waiting, _ := q.IsWaiting(w)
+		if !waiting {
+			t.Errorf("item %v was expected to be waiting but isn't", w)
+			errored = true
 		}
+	}
 
-		return false, nil
-	})
+	// check 'active' queue
+	if q.Len() != len(qs.active) {
+		t.Errorf("the active q has %d queued items but we expected %d", q.Len(), len(qs.active))
+		errored = true
+	}
+	// we can only check the active queued items when the queue implements the EnhancedInterface
+	if ei, ok := q.(EnhancedInterface); ok {
```

Comment:
Actually, it's the other way around, most of the time we use one of our constructors to create our implementation of the workqueue so it is nearly always an EnhancedInterface.  You actually have to plug in your own queue implementation but use our delaying queue to hit the path where it doesn't go down this path.

Reply:
After looking into this more, we will not pass the type check here because emebedded interface only promotes its methods but the hosting type (`delayingType`) does not fulfill `EnhancedInterface`. IIUC it cannot fulfill `EnhancedInterface` interface since it does not implement the methods as the promoted methods have different receivers.

You can test this by with ` _, ok := q.(ExtendedInterface)`) being false all the time. So the test here will always skip it.

other test `var _ ExtendedInterface = &delayingType{}` will complain on compile time